common/transport/amqp/src/amqp.ts

```diff
@@ -135,6 +88,280 @@ export class Amqp {
+          '*': () => this._fsm.deferUntilTransition()
+        },
+        connected: {
+          _onEnter: (connectCallback, result) => {
+            /*Codes_SRS_NODE_COMMON_AMQP_16_002: [The `connect` method shall establish a connection with the IoT hub instance and if given as argument call the `done` callback with a null error object in the case of success and a `results.Connected` object.]*/
+            this._safeCallback(connectCallback, null, new results.Connected(result));
+          },
+          amqpError: (err, callback) => this._fsm.transition('disconnecting', callback, err),
+          connect: (callback) => callback(null, new results.Connected()),
+          disconnect: (disconnectCallback) => {
+            this._fsm.transition('disconnecting', disconnectCallback);
+          },
+          initializeCBS: (callback) => {
+            this._cbs = new ClaimsBasedSecurityAgent(this._amqp);
+            this._cbs.attach(callback);
+          },
+          putToken: (audience, token, callback) => {
+            this._cbs.putToken(audience, token, callback);
+          },
+          send: (message: Message, endpoint: string, to: string, done: GenericAmqpBaseCallback<any>): void => {
+            /*Codes_SRS_NODE_COMMON_AMQP_16_006: [The `send` method shall construct an AMQP message using information supplied by the caller, as follows:
+            The ‘to’ field of the message should be set to the ‘to’ argument.
+            The ‘body’ of the message should be built using the message argument.] */
+            debug('call to deprecated api \'azure-iot-amqp-base.Amqp.send\'. You should be using SenderLink.send instead');
+            let amqpMessage = AmqpMessage.fromMessage(message);
+            if (to !== undefined) {
+              amqpMessage.properties.to = to;
+            }
+
+            if (!this._senders[endpoint]) {
+              this._fsm.handle('attachSenderLink', endpoint, null, (err) => {
+                if (err) {
+                  debug('failed to attach the sender link: ' + err.toString());
+                  done(err);
+                } else {
+                  (this._senders[endpoint] as SenderLink).send(amqpMessage, done);
+                }
+              });
+            } else {
+              (this._senders[endpoint] as SenderLink).send(amqpMessage, done);
+            }
+
+          },
+          getReceiver: (endpoint: string, done: GenericAmqpBaseCallback<ReceiverLink>): void => {
+            /*Codes_SRS_NODE_COMMON_AMQP_16_010: [If a receiver for this endpoint doesn’t exist, the getReceiver method should create a new AmqpReceiver object and then call the done() method with the object that was just created as an argument.] */
+            if (!this._receivers[endpoint]) {
+              this._fsm.handle('attachReceiverLink', endpoint, null, (err) => {
+                if (err) {
+                  done(err);
+                } else {
+                  done(undefined, this._receivers[endpoint]);
```

Comment:
use _safeCallback?  I assume done is option because you use _safeCallback below.

Reply:
callback is not optional here. 

---
In reply to: [139524981](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139524981) [](ancestors = 139524981)